Use Gitlab Container registry for CI

mkuron opened the pull request:
Depends on https://github.com/espressomd/docker/pull/26

mkuron committed:
Use Gitlab Container registry for CI

KaiSzuttor reviewed (changes requested):
You could use a global variable and prepend that to the current image names, that makes it easier to change the path in the future

mkuron commented:
@KaiSzuttor, we can't use variables there: https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/14332

KaiSzuttor commented:
I don't understand why...
https://docs.gitlab.com/ce/ci/yaml/README.html#variables

mkuron commented:
There is some new stuff in the EspressoScriptInterface that crashes the Intel compiler: https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/14526, somehow related to `boost::is_constructible`. No idea where that is used.

mkuron committed:
Intel 2015 fixes

boost::serialization::make_array in Boost < 1.64.0 is broken with Intel compilers at least up to 17.0.3
    https://svn.boost.org/trac10/ticket/12516
    https://svn.boost.org/trac10/ticket/12993
    https://github.com/boostorg/serialization/commit/c90e97eb1ed5e6f9ea3b1d3b5a825a9d2ef284b3

boost::mpi cannot be compiled in Boost 1.64
    https://svn.boost.org/trac10/ticket/12723
    https://github.com/boostorg/mpi/commit/00e55d8f5e5280f499bf34b2c41bb9b67c7c8549
    https://github.com/boostorg/mpi/commit/f5bdcc1ebfe954bb64835f2a0efd94471da42207

mkuron committed:
Intel MPI fix

mkuron committed:
Factory_test fix for Intel

If the derived class is empty, the object's typeid is that of the parent class. Thus dynamic_casting to the derived class fails.

mkuron commented:
Ready to merge.

KaiSzuttor commented on CMakeLists.txt:
```diff
@@ -292,10 +292,20 @@ if(WITH_TESTS)
 list(APPEND LIBRARIES ${Boost_LIBRARIES})
 
+if(Boost_VERSION VERSION_GREATER 106399 AND Boost_VERSION VERSION_LESS 106500)
+  message(FATAL_ERROR "Boost 1.64 is unsupported")
```
Why is Boost 1.64 not supported? It's currently the standard version in fedora...

mkuron replied:
Its Boost.MPI is incompatible with its Boost.Serialization. Boost 1.64 simply cannot be compiled if MPI is enabled. There is no way to make it work besides upgrading or downgrading. See https://svn.boost.org/trac10/ticket/12723 .

pkreissl replied:
Removing these lines, I can compile on my fedora 27, with boost version 1.64.
All tests run successfully (in parallel), except for `testsuite/thermalized_bond.py`, which only succeeds when run single-core.

mkuron replied:
Then Fedora must have backported that patch but left the Boost version unchanged. Can you check Fedora's source package repository for that? I guess we would need some way to detect whether we are using a vanilla Boost 1.64 or a patched one if we want to warn the user that their Boost version may be broken. If we don't do that, they will just get hard-to-understand error messages.
